### PR TITLE
Improve lode UI

### PR DESCRIPTION
**Pull request by sir-marc**
- Add some basic styles.
- Add triangle count
- Switch to one-pager
- Add LOD Viewer
image

### CODE REVIEW

File: lode-ui/src/LODViewer.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    return (
+      <Center h="100%">
+        <Spinner />
+      </Center>
+    );
+  }
+  return (
+    <Box h="100%">
+      <Canvas>
+        <OrbitControls autoRotate onWheel={() => console.log("update")} />
```
Comment (StuckiSimon) [inline lode-ui/src/LODViewer.js:52]: classic?

File: lode-ui/src/Model.js [hunk 2]
```diff
@@ -1,28 +1,71 @@
...
+
+function TriangleCount({ url }) {
+  const { gl } = useThree();
+  const [triangleCount, setTriangleCount] = useState(0);
+  useEffect(() => {
+    setTriangleCount(0);
+    setTimeout(() => {
+      setTriangleCount(gl.info.render.triangles);
+    }, 500);
+    // eslint-disable-next-line
```
Comment (StuckiSimon) [inline lode-ui/src/Model.js:30]: why?

File: lode-ui/src/Model.js [hunk 3]
```diff
@@ -1,28 +1,71 @@
...
+}
+
+function TriangleCount({ url }) {
+  const { gl } = useThree();
+  const [triangleCount, setTriangleCount] = useState(0);
+  useEffect(() => {
+    setTriangleCount(0);
+    setTimeout(() => {
+      setTriangleCount(gl.info.render.triangles);
+    }, 500);
```
Comment (StuckiSimon) [inline lode-ui/src/Model.js:29]: magic wait time 😅

File: lode-ui/package.json [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
   "dependencies": {
+    "@chakra-ui/react": "^1.6.0",
+    "@emotion/react": "^11",
+    "@emotion/styled": "^11",
     "@react-three/drei": "^4.3.2",
     "@react-three/fiber": "^6.0.19",
     "@testing-library/jest-dom": "^5.11.4",
     "@testing-library/react": "^11.1.0",
     "@testing-library/user-event": "^12.1.10",
+    "framer-motion": "^4",
```
Comment (StuckiSimon) [inline lode-ui/package.json:14]: I'd prefer having the same version ranges for @emotion and framer-motion as for all other dependencies - even if it's written like this in the Getting Started guide of Chakra 😄

File: lode-ui/src/App.js [hunk 5]
```diff
@@ -1,16 +1,23 @@
...
-      <Switch>
-        <Route path="/model/" component={ModelViewer} />
-        <Route path="/" component={ModelChooser} />
-      </Switch>
-    </Router>
+    <ChakraProvider>
+      <Router>
+        <Grid h="100%" templateColumns="200px 1fr" templateRows="1fr">
+          <GridItem borderWidth={1} rowSpan={2}>
+            <ModelChooser></ModelChooser>
```
Comment (StuckiSimon) [inline lode-ui/src/App.js:13]: why not self-closing?

File: lode-ui/src/App.js [hunk 6]
```diff
@@ -1,16 +1,23 @@
-import { BrowserRouter as Router, Route, Switch } from "react-router-dom";
-import "./App.css";
+import { BrowserRouter as Router } from "react-router-dom";
+import { ChakraProvider, Grid, GridItem } from "@chakra-ui/react";
 import ModelChooser from "./ModelChooser";
 import ModelViewer from "./ModelViewer";
+import "./App.css";
```
Comment (StuckiSimon) [inline lode-ui/src/App.js:5]: I'd suggest we use emotion consistently and remove the last remaining `.css` styling? 😄

**Comment by StuckiSimon** [comment]
Looks promising 👍 
I've added some comments.

File: lode-ui/src/InfoPanel.js [hunk 7]
```diff
@@ -3,26 +3,33 @@ import { Html } from "@react-three/drei";
...
+      justify="space-between"
+      align="flex-end"
+      direction="row"
       pointerEvents="none"
     >
       <Box bg="rgba(255, 255, 255, 0.5)" p={5}>
         Distance to object: {distance}
       </Box>
+      <Box bg="rgba(255, 255, 255, 0.5)" p={5}>
+        Current level: {currentLevel}
```
Comment (StuckiSimon) [inline lode-ui/src/InfoPanel.js:29]: Nice one, I barely noticed the transition 👍

File: lode-ui/src/Model.js (hunk 2 above)
Comment (sir-marc) [reply lode-ui/src/Model.js:30]: To not trigger the method when the gl object changes.

Comment (StuckiSimon) [reply lode-ui/src/Model.js:30]: I suggest to store it in a variable outside of the effect and use that variable for the effect change detection then. This way the setTimeout should actually not be required anymore and the whole code works nicer?

File: lode-ui/src/LodLevelDisplay.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { Box, Flex } from "@chakra-ui/layout";
+import { Html } from "@react-three/drei";
+import { useFrame } from "@react-three/fiber";
+import { useState } from "react";
+
+function LodLevelDisplay({ lod }) {
```
Comment (StuckiSimon) [inline lode-ui/src/LodLevelDisplay.js:6]: where is this component being used? 🤔

File: lode-ui/src/Model.js (hunk 2 above)
Comment (sir-marc) [reply lode-ui/src/Model.js:30]: Then there is no re-render to trigger the effect. Only other option i see would be to use the `useFrame` callback to update it, but seems like an overkill to calculate this every frame, as it is not supposed to change.

Comment (StuckiSimon) [reply lode-ui/src/Model.js:30]: How do you guarantee that the triangle count is correct after 500 ms?

Comment (sir-marc) [reply lode-ui/src/Model.js:30]: I dont :D I am aware of that.

Comment (StuckiSimon) [reply lode-ui/src/Model.js:30]: Fair - I don't think we need to over-optimize (as long as it works) 👍 
My only suggestion is to add a comment to answer the initial question "why?". This prevents future me or you of trying to improve it 😄 
something along the lines: "gl.info.render.triangles is a ref, therefore it cannot be used in effect, 500ms is sufficient for 99 percentile, whatever the reasoning"
In general: I try to add comments wherever I assume that someone with sufficient context might ask oneself "why? 🤔 "

File: cli/package.json [hunk 9]
```diff
@@ -21,6 +22,8 @@
     "gl-matrix": "^3.3.0",
     "glob": "^7.1.6",
     "inquirer": "^8.0.0",
+    "milliparsec": "^2.0.4",
+    "serve-static": "^1.14.1",
     "sirv": "^1.0.11",
```
Comment (StuckiSimon) [inline cli/package.json:27]: Is this still required?

File: cli/lib/server/index.js [hunk 10]
```diff
@@ -1,26 +1,73 @@
...
+    .get("/changes", (req, res) => {
+      const listener = (change) => {
+        const timestamp = new Date().getMilliseconds();
+        const manifest = readFile(
+          path.join(opts.outputFoldername, MANIFEST_FILENAME)
+        );
+        res
+          .status(200)
+          .send({ change, timestamp, manifest: JSON.parse(manifest) });
+        pubSub.unsub(listener);
```
Comment (StuckiSimon) [inline cli/lib/server/index.js:55]: this means that messages being sent in between will not be noticed by the client.
ok for me, as we're pragmatic.

File: cli/lib/server/index.js [hunk 11]
```diff
@@ -1,26 +1,73 @@
...
+  const pubSub = createPubSub();
 
   app
-    .use(function someMiddleware(req, res, next) {
-      res.setHeader("Access-Control-Allow-Origin", "*");
-      next();
-    })
-    .use("/assets", sirv(outputFoldername))
+    .use(json())
+    .use(cors({ origin: "*", allowedHeaders: "*" }))
```
Comment (StuckiSimon) [inline cli/lib/server/index.js:34]: nice

**Comment by StuckiSimon** [comment]
image
Exceptions are fine, but please provide a comment on the reasoning. This helps - especially as we have no tests so far.

**Pull request merged**